Initial gateware code for VGA capture

rohitk-singh opened the pull request:
Changes:
- Added gateware code for vga capture in `gateware/vga`
- Added `atlys_vga.py` target for vga capture
- Modified atlys platform file with vga pins
- Added some firmware code for vga capure in `ad9984a.c` and `vga_in.c`

Doesn't have to merged compulsorily. If any suggestions or refactoring/restructuring ideas are there, they are welcome. Any comments also welcome.

Have to add rest of the firmware code for capture to work. There are some conflicts with the code, so haven't committed them yet. Looking into them right now

mithro commented on gateware/vga/testcapture.py:
Can you please fix the wrapping here.

Something like;
```python
    print("cycle: {cycle} , hsync: {hsync}, vsync: {vsync}, de: {de}, r: {r}, g: {g}, b: {b}".format(
        cycle = selfp.simulator.cycle_counter,
        hsync = selfp.hsync,
        vsync = selfp.vsync,
        de = selfp.de,
        r = selfp.r,
        g = selfp.g,
        b = selfp.b,
        ))
```

mithro commented on firmware/lm32/ad9984a.c:
```diff
@@ -0,0 +1,94 @@
+#include "ad9984a.h"
+
+
+const unsigned char regReadAddress[NUM_REG_READ] = {0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 
```
Can you add some comments about what is going on here?

mithro commented on firmware/lm32/ad9984a.c:
```diff
@@ -0,0 +1,94 @@
+#include <stdio.h>
```
Can you add some comments linking to were datasheets for this part can be found.

It would also be good to add some comments about the VGA board here.

mithro commented on firmware/lm32/vga_in.c:
```diff
@@ -0,0 +1,198 @@
+	mask &= ~(1 << VGA_IN_INTERRUPT);
+	irq_setmask(mask);
+
+    vga_in_frame_start_counter_write(0); // stop counter
```
Formatting..

mithro commented:
Hi @rohit91,

Generally things are looking pretty good but there are some small formatting issues. Can you please do the following things;
 * Run the python code through pep8. Due to the way migen code is generally formatted, you won't be able to be 100% pep8 clean, but it should at least show you where you have some formatting problems.
 * Fix tabs/spaces in the C code. It seems most of the existing code uses tabs.

mithro commented:
@enjoy-digital Can you take a close look at the migen code?

mithro commented:
@rohit91 Was this the code that @CarlFK got working today?

mithro commented on firmware/lm32/ad9984a.c:
```diff
@@ -0,0 +1,107 @@
+ * www.analog.com/media/en/technical-documentation/data-sheets/AD9984A.pdf
+ * 
+ * Code below is for AD9984A based VGA capturing for HDMI2USB using VGA 
+ * Expansion Board for Atlys(https://github.com/timvideos/HDMI2USB-vmodvga)
```
Space between Atlys and (

mithro commented on firmware/lm32/ad9984a.c:
```diff
@@ -0,0 +1,107 @@
+ * Analog Devices' AD9984A VGA Capture IC. 
+ * 
+ * Refer the AD9984A datasheet for Register details here:
+ * www.analog.com/media/en/technical-documentation/data-sheets/AD9984A.pdf
```
Add http:// in front of www

mithro commented on firmware/lm32/ad9984a.c:
```diff
@@ -0,0 +1,107 @@
+#include "ad9984a.h"
+
+/* Array of AD9984A register addresses to read values from */
+const unsigned char regReadAddress[NUM_REG_READ] = {0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 
```
What do these numbers mean? Can you please document them?

mithro commented on firmware/lm32/ad9984a.c:
```diff
@@ -0,0 +1,107 @@
+
+unsigned char ad9984a_read(unsigned char reg)
+{
+    unsigned char chip_addr = AD9984A;  // 7-bit, Other possible adr=0x4D
```
This should probably be a "byte" or "uint8_t" right?

mithro commented on firmware/lm32/vga_in.c:
```diff
@@ -0,0 +1,197 @@
+static int vga_in_next_fb_index;
+static int vga_in_hres, vga_in_vres;
+
+void vga_in_isr(void)
```
Is this code identical to the HDMI version? Or does it have some customisations?

mithro commented on firmware/lm32/vga_in.c:
```diff
@@ -0,0 +1,197 @@
+    printf("\nDumping vga_in fb data and registers\n");
+    int i;
+    unsigned int upper=0, lower=0;
+    flush_l2_cache();
```
Why do you have to flush the l2 cache here? Can you please add a comment?

mithro commented on firmware/lm32/vga_in.c:
```diff
@@ -0,0 +1,197 @@
+    int i;
+    unsigned int upper=0, lower=0;
+    flush_l2_cache();
+    volatile unsigned int *framebuffer = (unsigned int *)(MAIN_RAM_BASE + VGA_IN_FRAMEBUFFERS_BASE);
```
This should probably be something like uint32_t* ?

mithro commented on gateware/vga/dma.py:
```diff
@@ -0,0 +1,134 @@
+from migen.fhdl.std import *
```
Should we add a comment about trying to share the DMA between the HDMI and VGA code? Something like;
```c
/* TODO: Investigate how to share the DMA code between HDMI (see gateware/hdmi/dma.py) and VGA code. See https://github.com/timvideos/HDMI2USB-misoc-firmware/issues/333. */
```

rohit91 replied:
Sigh, how did I miss that. We are using DMA from `gateware/hdmi_in/dma.py` itself. This file probably got copied initially and was never removed. 
![screenshot280](https://cloud.githubusercontent.com/assets/1108318/14052233/9dccfe50-f2ee-11e5-9ca1-993a442a86aa.png)

mithro replied:
That is why we do code review :)

mithro commented on gateware/vga/testcapture.py:
```diff
@@ -0,0 +1,199 @@
+from migen.fhdl.std import *
```
This is a test bench right? Can you add a python docstring at the top of this file?

mithro commented on platforms/atlys.py:
```diff
@@ -441,6 +441,23 @@
 #            Subsignal("tck", Pins("A15")),
 #        ),
 
```
**Don't** think we should block merging on this, but I think we need to figure out how to the VHDCI as a connector.

See https://github.com/m-labs/migen/blob/legacy/mibuild/platforms/papilio_pro.py#L45 and https://github.com/m-labs/migen/blob/11f1b134a9165513d32258cda2bae1e1d590a020/migen/build/platforms/pipistrello.py#L116

Dunno how you "use" a connector in migen/misoc though... @enjoy-digital Do you know how?

mithro commented on targets/atlys_vga.py:
```diff
@@ -0,0 +1,241 @@
+# Support for the Digilent Atlys board - digilentinc.com/atlys/
```
I think this file needs to be updated for the latest changes in the other atlys targets? `_get_firmware_data` has been removed.

mithro commented on targets/atlys_vga.py:
```diff
@@ -0,0 +1,241 @@
+""", sys_clk=self.crg.cd_sys.clk)
+
+
+class MiniSoC(BaseSoC):
```
I thought that the having both Ethernet and the VGA board running at the same time wasn't possible due to IO level incompatibilities?

mithro commented on targets/atlys_vga.py:
```diff
@@ -0,0 +1,241 @@
+# Support for the Digilent Atlys board - digilentinc.com/atlys/
+import struct
```
Can you import most of the stuff in this file from the `atlys_base.py` file? (Like what the `atlys_hdmi2usb.py` does?)

rohit91 commented:
Starting afresh. 

Changes:
- Initial gateware code for vga capture in `gateware/vga`
- Modified `platforms/atlys.py` platform file by adding vga signals

@mithro Can you look at this and see if it can be merged or not. pep8 checks are ok (only visual indentation warnings and few >79 line-length warnings). I have incorporated all of @enjoy-digital's suggestions he had after his review. This gateware code is all that is necessary for vga capture. I'll be creating separate pull requests each for firmware and for target.

mithro commented on gateware/vga/datacapture.py:
```diff
@@ -0,0 +1,92 @@
+               self.counterY.eq(0),
+            ),
+
+            #  VGA Scan Timing Values used below for 1024x768@60Hz
```
Can you add a TODO to make this configurable / controllable?

mithro commented on gateware/vga/analysis.py:
```diff
@@ -0,0 +1,146 @@
+from gateware.csc.ycbcr444to422 import YCbCr444to422
+
+
+class FrameExtraction(Module, AutoCSR):
```
Can you add a TODO for tests around this?

mithro commented on gateware/vga/datacapture.py:
```diff
@@ -0,0 +1,92 @@
+from migen.bank.description import *
+
+
+class DataCapture(Module, AutoCSR):
```
A quick docstring comment here describing what this is doing would be useful.

mithro commented on gateware/vga/__init__.py:
```diff
@@ -0,0 +1,35 @@
+from gateware.vga.datacapture import DataCapture
+
+
+class VGAIn(Module, AutoCSR):
```
This is VGAIn from an AD9938 chip?

rohit91 replied:
yes. *AD9984A

rohitk-singh committed:
Initial gateware code for VGA capture

Files:
    new file:   gateware/vga/__init__.py
    new file:   gateware/vga/analysis.py
    new file:   gateware/vga/datacapture.py

__init__.py:
    Implements VGAIn module which instantiates submodules
    Datacapture, FrameExtrantion and DMA, and connects them

analysis.py:
    Implements FrameExtraction module, which is reponsible
    for sof(start of frame) detection, color space conversion,
    framing(packing) and also uses async fifo to move data
    from VGA pixel clock domain to sys_clk domain

datacapture.py:
    Implements DataCapture module which is responsible for
    capturing pixel data at proper time, depending on HSYNC
    and VSYNC signals

Currently only supports 1024x768@60Hz resolution capture

rohitk-singh committed:
Modified atlys_video target to include VGAIn module also.

VGAIn modules gives VGA capture functionality to atlys.
Also, we are now inheriting BaseSoC class directly instead of
MinoSoC

rohitk-singh committed:
Modified atlys platform file to add VGA signals.

Added to _io_vccb2_3v3 list because currently VGA signals are driven
at 3.3V. This is going to change and become configurable later.

rohitk-singh committed:
Fix docstring style to be compliant with migen docstring style

rohitk-singh committed:
Fix minor issues in VGAIn DataCapture module
